app/src/test/java/com/farmerbb/taskbar/ui/TaskbarControllerTest.java

```diff
@@ -0,0 +1,220 @@
+package com.farmerbb.taskbar.ui;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.graphics.Color;
+import android.view.Gravity;
+import android.widget.ImageView;
+
+import androidx.test.core.app.ApplicationProvider;
+
+import com.farmerbb.taskbar.R;
+import com.farmerbb.taskbar.util.U;
+
+import org.bouncycastle.jcajce.provider.symmetric.Grain128;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PowerMockIgnore;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.rule.PowerMockRule;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.android.controller.ServiceController;
+
+import static com.farmerbb.taskbar.util.Constants.POSITION_BOTTOM_LEFT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_BOTTOM_RIGHT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_BOTTOM_VERTICAL_LEFT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_BOTTOM_VERTICAL_RIGHT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_TOP_LEFT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_TOP_RIGHT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_TOP_VERTICAL_LEFT;
+import static com.farmerbb.taskbar.util.Constants.POSITION_TOP_VERTICAL_RIGHT;
+import static com.farmerbb.taskbar.util.Constants.PREF_START_BUTTON_IMAGE;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.robolectric.util.ReflectionHelpers.ClassParameter.from;
+import static org.robolectric.util.ReflectionHelpers.callInstanceMethod;
+
+@RunWith(RobolectricTestRunner.class)
+@PowerMockIgnore({"org.mockito.*", "org.robolectric.*", "android.*", "androidx.*"})
+@PrepareForTest(U.class)
+public class TaskbarControllerTest {
+    @Rule
+    public PowerMockRule rule = new PowerMockRule();
+
+    private ServiceController<TaskbarUIHostService> controller;
+    private TaskbarUIHostService hostService;
+    private TaskbarController uiController;
+    private Context context;
+    SharedPreferences prefs;
+
+    @Before
+    public void setUp() {
+        controller = Robolectric.buildService(TaskbarUIHostService.class);
+        hostService = controller.create().get();
+        uiController = hostService.controller;
```

Comment:
Any reason why we couldn't just construct a new `TaskbarController` here instead of wrapping it inside of a mock `UIHostService`?

Reply:
Actually, nevermind... There's enough logic inside of `UIHostService` that's good to have some coverage for

Reply:
The `TaskbarController` needs `UIHostService`, and its important logic needs `UIHostService` as parameter. An `UIHostService` instance is needed for testing `TaskbarController`. So I create a mock `UIHostService`, and start it as a normal service to test `TaskbarController` with more real occasion.

Reply:
I just realized we've already got `UIHostServiceTest` for testing `UIHostService`'s logic.

In that case, there's no reason to have the `TaskbarUIHostService` in the first place.  `TaskbarController` does not directly depend on `UIHostService` for any of its functionality, it only needs some form of `UIHost` that it can use to add / remove views to.  If you have plans to test the `drawTaskbar()` function itself, or any of the `UIController` callbacks, I think just a lightweight mock version of `UIHost` would suffice.

For what it's worth, the current tests still pass with `TaskbarUIHostService` deleted and the `setUp()` method of `TaskbarControllerTest` changed to the following:

```
    @Before
    public void setUp() {
        context = ApplicationProvider.getApplicationContext();
        uiController = new TaskbarController(context);
        prefs = U.getSharedPreferences(context);
    }
```